kernel/src/process/process_vm/mod.rs

```diff
@@ -175,47 +153,46 @@ impl ProcessVm {
         envp: Vec<CString>,
         aux_vec: AuxVec,
     ) -> Result<()> {
-        let root_vmar: ProcessVmarGuard<'_> = self.lock_root_vmar();
-        self.init_stack
-            .map_and_write(root_vmar.unwrap(), argv, envp, aux_vec)
+        let vmar = self.vmar.as_ref().unwrap();
+        self.inner
+            .init_stack
+            .map_and_write(vmar, argv, envp, aux_vec)
     }
 
-    pub(super) fn heap(&self) -> &Heap {
-        &self.heap
+    pub fn heap(&self) -> &Heap {
+        &self.inner.heap
     }
 
     #[cfg(target_arch = "riscv64")]
     pub(super) fn vdso_base(&self) -> Vaddr {
-        self.vdso_base.load(Ordering::Relaxed)
+        self.inner.vdso_base.load(Ordering::Relaxed)
     }
 
     #[cfg(target_arch = "riscv64")]
     pub(super) fn set_vdso_base(&self, addr: Vaddr) {
-        self.vdso_base.store(addr, Ordering::Relaxed);
+        self.inner.vdso_base.store(addr, Ordering::Relaxed);
     }
 
     /// Clears existing mappings and then maps the heap VMO to the current VMAR.
-    pub fn clear_and_map(&self) {
-        let root_vmar = self.lock_root_vmar();
-        root_vmar.unwrap().clear().unwrap();
-        self.heap.alloc_and_map_vm(root_vmar.unwrap()).unwrap();
+    pub(super) fn clear_and_map_heap(&self) {
+        let vmar = self.vmar().unwrap();
+        vmar.clear().unwrap();
+        self.inner.heap.alloc_and_map_vm(vmar).unwrap();
     }
 }
 
-/// Renews the [`ProcessVm`] of of the current process and then maps the heap VMO to the new VMAR.
-pub(super) fn renew_vmar_and_map_heap(ctx: &Context) {
-    let process_vm = ctx.process.vm();
-    let mut root_vmar = process_vm.lock_root_vmar();
+/// Unshares and renews the [`ProcessVm`] of of the current process.
+pub(super) fn unshare_and_renew_vm(ctx: &Context) {
+    let mut process_vm = ctx.process.vm().lock();
```

Comment:
Perhaps we should keep this lock until `ProgramToLoad::load_to_vm`. This will prevent other processes from seeing an inconsistent state during `execve` via `procfs`.